### DISCUSSION TITLE
Add an (U)Int128 type

### DISCUSSION THREAD

**Discussion by luizfernandonb** [dotnet/csharplang#4839] [Language Ideas]
**Moved from [this ](https://github.com/dotnet/runtime/issues/20640)issue from dotnet/runtime**

Proposal:

> While I know it does exist BigInteger sometimes it would be needed to have a type that is explicit 128 bit and nothing more probably it would be more efficient too (at least in a 64 bit machine at it could be represented as a couple of longs).
> 
> An Int128 is the "natural" type of GUID, IP6 addresses and more at low level but nevertheless true x86 instruction CPUID returns 4 integers (in 4 registers) that need to be "concatenated" to get the real value, the real "intent" was to return an Int128.
> 
> I have doubt if an "alias" could exist if a long is an abbreviation of "longword" and Int128 could be octa as abbreviation of "octaword"? The literal at that point could be "O" or "o"...

Proposal from me:

I think a `long long` as it is in C/C++ for would be useful to create a 128-bit integer in C#, since just the `long` keyword in C# already creates a 64-bit integer

**Comment by huoyaoyuan** [comment]
> I think a `long long` as it is in C/C++ for would be useful to create a 128-bit integer in C#, since just the `long` keyword in C# already creates a 64-bit integer

No. This will add confusion and complexity of language parser. **NO** keyword should be added for this type. `Int128` is even more brief.

(The case for `nint` was different, because it exposes some raw IL instructions which are not available in `IntPtr` metadata.)

The only thing requiring compiler support should be creating literals.

**Reply by luizfernandonb** [reply]
> No. This will add confusion and complexity of language parser. **NO** keyword should be added for this type. `Int128` is even more brief.

If following this line, then the `int` and `long` keywords should not exist, as they would add a complication to the language parser

But since there can't be a `long long`, maybe there could be a `llong`? The same way there is `nint`, or since it can't exist anyway, the most viable would then be to use the `Int128` structure

**Reply by aaronfranke** [reply]
Agreed, `Int128` is fine, and is shorter than `long long`, and barely longer than `llong` (which isn't super readable as it can be mistaken for `long`. `longer` would be more distinct from `long` than `llong` and is one word unlike `long long`, but it's the same number of characters as `Int128` anyway.

**Reply by jnm2** [reply]
> If following this line, then the `int` and `long` keywords should not exist, as they would add a complication to the language parser

@luizfernandonb I don't think that was the point. The point about complexity of language parsing was that no predefined types in C# consist of multiple keywords separated by whitespace.

**Reply by AraHaan** [reply]
Agreed, I think what we need is a proposal that everyone agrees with with a properly crafted keyword.

hmm how about:
``int128`` (System.Int128)
``uint128`` (System.UInt128)

That way it avoids parser complexity and also readable too. The downside is that it's all lowercase version of the typename but the benefit is that then they do not have to type ``System.Int128`` or ``System.UInt128``.

**Reply by luizfernandonb** [reply]
> > If following this line, then the `int` and `long` keywords should not exist, as they would add a complication to the language parser
> 
> @luizfernandonb I don't think that was the point. The point about complexity of language parsing was that no predefined types in C# consist of multiple keywords separated by whitespace.

Ah ok, now I understand what @huoyaoyuan meant

**Reply by luizfernandonb** [reply]
> Agreed, I think what we need is a proposal that everyone agrees with with a properly crafted keyword.
> 
> hmm how about:
> `int128` (System.Int128)
> `uint128` (System.UInt128)
> 
> That way it avoids parser complexity and also readable too. The downside is that it's all lowercase version of the typename but the benefit is that then they do not have to type `System.Int128` or `System.UInt128`.

Regardless of the type name, the only certainty we have is that it will be considerably faster than BigInteger, any native integer type is considerably faster than it

**Reply by tannergooding** [reply]
> The only thing requiring compiler support should be creating literals.

Its worth noting that its not just literals. The main blocker for the framework/runtime exposing anything is the ability to define checked variants of the operators. This would be unblocked by https://github.com/dotnet/csharplang/issues/4665.

Other considerations, however, include (but aren't necessarily limited to):
* literals
* constants
* constant folding
* operators resolve and behave differently for language built-in vs user-defined

**Reply by foxesknow** [reply]
> 
> 
> > If following this line, then the `int` and `long` keywords should not exist, as they would add a complication to the language parser
> 
> @luizfernandonb I don't think that was the point. The point about complexity of language parsing was that no predefined types in C# consist of multiple keywords separated by whitespace.

Well, there's `yield return` and `yield break`...

**Reply by spydacarnage** [reply]
Neither of which are _predefined types_...

**Comment by Atomosk** [comment]
LMAO, and `u512` would be `long long long ulong` or `very long ulong`? I hope you are just joking.

**Reply by luizfernandonb** [reply]
`long long` is no longer possible, see @huoyaoyuan comment, I've already changed my mind

**Reply by yaakov-h** [reply]
Personally, I’d hope for 512 to be `longboi`

**Reply by Unknown6656** [reply]
> Personally, I’d hope for 512 to be `longboi`

@333fred   I'll be angry and extremely disappointed if this is not implemented in the next iteration of C#. I need `longboi` as a keyword in my daily development workflow!

**Reply by 333fred** [reply]
Uh huh.

**Comment by ufcpp** [comment]
> I think a long long as it is in C/C++ for would be useful

https://www.reddit.com/r/linux/comments/5f5rfd/this_is_why_linus_doesnt_accept_prs_from_github/

**Reply by luizfernandonb** [reply]
LMAO 🤣🤣🤣🤣🤣🤣🤣

**Comment by deeprobin** [comment]
how about `overlong`, `extendedlong` or `loong` (double o)

**Reply by kyoyama-kazusa** [reply]
```csharp
loong a = 1;
dragon b = 2;
```

🤣

**Reply by jnm2** [reply]
`overextended long`

**Reply by deeprobin** [reply]
```
long = Int64
loong = Int128
looong = Int256
loooong = Int512
```

**Reply by AraHaan** [reply]
How about having an int type that does not have a fixed bitness so the max value is infinity and min value is negative infinity?

(and to represent that build in the infinity symbol into the language for the min and max value for it)

**Reply by Atomosk** [reply]
> How about having an int type that does not have a fixed bitness so the max value is infinity and min value is negative infinity?
> 
> (and to represent that build in the infinity symbol into the language for the min and max value for it)

Now that might become the first real Turing complete language. With infinite memory tape.

**Reply by Xyncgas** [reply]
We can define such type, type is just a set, now to instantiate this type yes let the compiler figure out what's the extend of your shortness and longness and the binary format for these values will be in 'compressed7BitInteger' or https://arxiv.org/pdf/1209.2137.pdf
dynamicInt is the best idea

**Reply by iam3yal** [reply]
`long2`, `long4`, `long8`. 😄

**Comment by silverf1re** [comment]
I don't care what its called but .net 7 support uint128 and int128 so the C# team should figure out what they want to call it and implement something so we can use it. https://devblogs.microsoft.com/dotnet/dotnet-7-generic-math/#types-without-language-support

**Reply by Xyncgas** [reply]
These programmers are the back bone of our industry, we make money off their work with our business, so let's give them some money and let's get after it and let's cut them some slacks because we all should just have a good time. Until Russia's nuclear war in Feb/28 https://banned.video

**Reply by iam3yal** [reply]
@silverf1re The people that need these types are using them today, you can either wait or use them too. 😄
